Framework/Crystal/src/PredictSatellitePeaks.cpp

```diff
@@ -382,36 +375,96 @@ void PredictSatellitePeaks::predictOffsetsWithCrossTerms(V3D offsets1, V3D offse
-        vector<int> SavPk{RunNumber, boost::math::iround(1000.0 * satelliteHKL[0]),
-                          boost::math::iround(1000.0 * satelliteHKL[1]), boost::math::iround(1000.0 * satelliteHKL[2])};
+    // Check if Q is non-physical
+    if (Qs.Z() * m_qConventionFactor <= 0)
+      return nullptr;
 
-        bool foundPeak = binary_search(AlreadyDonePeaks.begin(), AlreadyDonePeaks.end(), SavPk);
+    std::shared_ptr<IPeak> satellite_iPeak = Peaks->createPeak(Qs, 1);
 
-        if (!foundPeak) {
-          AlreadyDonePeaks.emplace_back(SavPk);
-        } else {
-          continue;
-        }
+    return satellite_iPeak;
+  }
 
-        peak->setHKL(satelliteHKL * m_qConventionFactor);
-        peak->setIntHKL(hkl * m_qConventionFactor);
-        peak->setRunNumber(RunNumber);
-        peak->setIntMNP(V3D(m, n, p) * m_qConventionFactor);
-        outPeaks->addPeak(*peak);
-      }
+  else if (workspace_type == workspace_type_enum::lean_elastic_peaks) {
+    Kernel::V3D const Qs = UB * satelliteHKL * 2.0 * M_PI * m_qConventionFactor;
+
+    std::shared_ptr<IPeak> satellite_iPeak = Peaks->createPeakQSample(Qs);
+
+    return satellite_iPeak;
+  }
+
+  else
+    return nullptr;
+}
+
+void PredictSatellitePeaks::addPeakToOutputWorkspace(std::shared_ptr<IPeak> satellite_iPeak,
+                                                     Kernel::Matrix<double> const &peak_goniometer_matrix,
+                                                     Kernel::V3D const &hkl, Kernel::V3D const &satelliteHKL,
+                                                     int const RunNumber,
+                                                     std::vector<std::vector<int>> &AlreadyDonePeaks,
+                                                     Kernel::V3D const &mnp) {
+  if (satellite_iPeak == nullptr)
+    return;
+
+  workspace_type_enum const workspace_type = determine_workspace_type(Peaks);
+
+  if (workspace_type == workspace_type_enum::regular_peaks) {
+    Geometry::InstrumentRayTracer const tracer(Peaks->getInstrument());
+
+    std::shared_ptr<Peak> const peak = std::dynamic_pointer_cast<Peak>(satellite_iPeak);
+
+    if (!peak->findDetector(tracer))
+      return;
+  }
+
+  std::vector<int> const SavPk{RunNumber, boost::math::iround(1000.0 * satelliteHKL[0]),
+                               boost::math::iround(1000.0 * satelliteHKL[1]),
+                               boost::math::iround(1000.0 * satelliteHKL[2])};
+
+  bool const foundPeak = binary_search(AlreadyDonePeaks.begin(), AlreadyDonePeaks.end(), SavPk);
+
+  if (!foundPeak) {
+    AlreadyDonePeaks.emplace_back(SavPk);
+  }
+
+  else
+    return;
+
+  satellite_iPeak->setGoniometerMatrix(peak_goniometer_matrix);
+  satellite_iPeak->setHKL(satelliteHKL * m_qConventionFactor);
+  satellite_iPeak->setIntHKL(hkl * m_qConventionFactor);
+  satellite_iPeak->setRunNumber(RunNumber);
+  satellite_iPeak->setIntMNP(mnp * m_qConventionFactor);
+
+  if (workspace_type == workspace_type_enum::regular_peaks) {
```

Comment:
You shouldn't need to do this if/else casting since the peak workspaces have `addPeak(const Geometry::IPeak &ipeak)`, it will then cast to the correct type.

Just do
`outPeaks->addPeak(*satellite_iPeak);`

Reply:
done!